### PR TITLE
feat: Identifier can be customized for sensitive surfaces

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR reduces the width of the sensitive part of the geometry identifier from 28bits to 20bits, to make room for an 8bit *extra* identifier part. The idea is that this can be used by experiments to label sensitive elements. The bit masks after the change look like:

```cpp
static constexpr Value kVolumeMask    = 0xff00000000000000; // (2^8)-1 = 255 volumes
static constexpr Value kBoundaryMask  = 0x00ff000000000000; // (2^8)-1 = 255 boundaries
static constexpr Value kLayerMask     = 0x0000fff000000000; // (2^12)-1 = 4096 layers
static constexpr Value kApproachMask  = 0x0000000ff0000000; // (2^8)-1 = 255 approach surfaces
static constexpr Value kSensitiveMask = 0x000000000fffff00; // (2^20)-1 sensitive surfaces
static constexpr Value kExtraMask     = 0x00000000000000ff; // (2^8)-1 extra values
```

In addition, it adds a hook function to the geometry closure of signature

```cpp
GeometryIdentifier(GeometryIdentifier orig, const Surface& surface);
```

to customize the geometry identifier. Technically, this hook can be used to also reset the volume and layer ids, but is primarily intended to set the *extra* components.

### CODE REVIEW

File: Core/include/Acts/Geometry/GeometryIdentifier.hpp [hunk 1]
```diff
@@ -74,18 +78,20 @@ class GeometryIdentifier {
...
-  static constexpr Value kApproachMask = 0x0000000ff0000000;
-  // (2^28)-1 sensitive surfaces
-  static constexpr Value kSensitiveMask = 0x000000000fffffff;
+  // clang-format off
+  static constexpr Value kVolumeMask    = 0xff00000000000000; // (2^8)-1 = 255 volumes
+  static constexpr Value kBoundaryMask  = 0x00ff000000000000; // (2^8)-1 = 255 boundaries
+  static constexpr Value kLayerMask     = 0x0000fff000000000; // (2^12)-1 = 4096 layers
+  static constexpr Value kApproachMask  = 0x0000000ff0000000; // (2^8)-1 = 255 approach surfaces
+  static constexpr Value kSensitiveMask = 0x000000000fffff00; // (2^20)-1 sensitive surfaces
+  static constexpr Value kExtraMask     = 0x00000000000000ff; // (2^8)-1 extra values
```
Comment (andiwand) [inline Core/include/Acts/Geometry/GeometryIdentifier.hpp:93]: should we add the explicit number of combinations here too?

```suggestion
  static constexpr Value kSensitiveMask = 0x000000000fffff00; // (2^20)-1 = 1048575 sensitive surfaces
  static constexpr Value kExtraMask     = 0x00000000000000ff; // (2^8)-1 = 255 extra values
```

File: Core/include/Acts/Geometry/GeometryIdentifier.hpp [hunk 2]
```diff
@@ -124,6 +130,9 @@ class GeometryIdentifier {
 
 std::ostream& operator<<(std::ostream& os, GeometryIdentifier id);
 
+using GeometryIdentifierHook =
+    std::function<GeometryIdentifier(GeometryIdentifier, const Surface&)>;
```
Comment (andiwand) [inline Core/include/Acts/Geometry/GeometryIdentifier.hpp:134]: I like it :+1: apart from changing the identifier it should also be easy to build a mapping between acts and experiment IDs

File: Core/include/Acts/Geometry/GeometryIdentifier.hpp [hunk 3]
```diff
@@ -53,6 +55,8 @@ class GeometryIdentifier {
   constexpr Value approach() const { return getBits(kApproachMask); }
   /// Return the sensitive identifier.
   constexpr Value sensitive() const { return getBits(kSensitiveMask); }
+  /// Return the extra identifier
+  constexpr Value extra() const { return getBits(kExtraMask); }
```
Comment (andiwand) [inline Core/include/Acts/Geometry/GeometryIdentifier.hpp:59]: should we state possible use cases somewhere?

one thing that comes to my mind is tagging surfaces of interest e.g. something in the endcaps that can easily be filtered while construction

File: Tests/UnitTests/Core/Geometry/TrackingGeometryClosureTests.cpp [hunk 4]
```diff
@@ -140,9 +151,113 @@ BOOST_AUTO_TEST_CASE(GeometryIdentifier_closeGeometry_test) {
...
+      GeometryIdentifier::Value layer_id = 0;
+      for (auto lay : vol.confinedLayers()->arrayObjects()) {
+        // check the layer volume id and layer layer id
+        auto lay_vol_id = lay->geometryId().volume();
+        auto lay_lay_id = lay->geometryId().layer();
+        BOOST_CHECK_EQUAL(++layer_id, lay_lay_id);
+        BOOST_CHECK_EQUAL(geoid, lay_vol_id);
+        // test the layer approach surfaces
+        if (lay->approachDescriptor() != nullptr) {
+          // approach surfacesare counted from 1 - n
```
Comment (andiwand) [inline Tests/UnitTests/Core/Geometry/TrackingGeometryClosureTests.cpp:229]:
```suggestion
          // approach surfaces are counted from 1 - n
```

**Review by andiwand — APPROVED** [review]
I like it :+1:

**Comment by asalzburger** [comment]
So, if we had lets say 20 surfaces front size and 20 surfaces backside, would you do 0-20 and a side bit, but 0-40 and a side bit ?

File: Core/include/Acts/Geometry/GeometryIdentifier.hpp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/Geometry/GeometryIdentifier.hpp:93]: Sure, before it was hundreds of millions, so I didn't bother

File: Core/include/Acts/Geometry/GeometryIdentifier.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/GeometryIdentifier.hpp:134]: > So, if we had lets say 20 surfaces front size and 20 surfaces backside, would you do 0-20 and a side bit, but 0-40 and a side bit ?

Answered in person: second one will keep the current behaviour ... Good.

**Review by asalzburger — APPROVED** [review]
Good thing.

**Comment by paulgessinger** [comment]
I addressed @andiwand's comments. Can one of you approve again?

**Review by andiwand — APPROVED** [review]

**Comment by paulgessinger** [comment]
Hmm. I didn't see hash failures locally...

**Comment by paulgessinger** [comment]
The fast-test right after FATRAS only already fails. I'll have to investigate. I'm taking this WIP again.

File: Core/include/Acts/Geometry/GeometryIdentifier.hpp [hunk 5]
```diff
@@ -74,18 +80,20 @@ class GeometryIdentifier {
...
-  // (2^12)-1 = 4096 layers
-  static constexpr Value kLayerMask = 0x0000fff000000000;
-  // (2^8)-1 = 255 approach surfaces
-  static constexpr Value kApproachMask = 0x0000000ff0000000;
-  // (2^28)-1 sensitive surfaces
-  static constexpr Value kSensitiveMask = 0x000000000fffffff;
+  // clang-format off
+  static constexpr Value kVolumeMask    = 0xff00000000000000; // (2^8)-1 = 255 volumes
+  static constexpr Value kBoundaryMask  = 0x00ff000000000000; // (2^8)-1 = 255 boundaries
+  static constexpr Value kLayerMask     = 0x0000fff000000000; // (2^12)-1 = 4096 layers
```
Comment (AJPfleger) [inline Core/include/Acts/Geometry/GeometryIdentifier.hpp:92]:
```suggestion
  static constexpr Value kLayerMask     = 0x0000fff000000000; // (2^12)-1 = 4095 layers
```

**Comment by paulgessinger** [comment]
Ok I think the hash changes come from the fact that we write the actual geometry id 64bit integer value to the outputs, which clearly changes if we reduce the bit width of the sensitive id.

**Review by andiwand — APPROVED** [review]

**Pull request merged**